core/bootstrap/src/main/java/org/wildfly/swarm/bootstrap/util/JarFileManager.java

```diff
@@ -0,0 +1,39 @@
+package org.wildfly.swarm.bootstrap.util;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.jar.JarFile;
+
+public class JarFileManager {
+
+    public static final JarFileManager INSTANCE = new JarFileManager();
+
+    private JarFileManager() {
+    }
+
+    public JarFile addJarFile(File file) throws IOException {
+
+        JarFile jarFile = jarFileToClose.get(file);
+        if (jarFile == null) {
+             jarFile = new JarFile(file);
```

Comment:
So we're creating a `new JarFile` here so that we can `close` it later? That sounds strange, but if it helps... At least a comment would be welcome.

Reply:
Aha, we also `return` it. OK, nevermind :-)

Reply:
Yes that's exactly the key here. JarFile should be closed before they can be removed in Windows. Maybe a comment in the class instead of the method?

Reply:
Ok

Reply:
@Ladicek pushed suggested change